Intro tidyverse learnr

GilHenriques opened the pull request:
This is the first draft of the Introduction to the Tidyverse learnr tutorial.
It is just a first draft so there are probably plenty of typos and, certainly, much room for improvement.
Any suggestions are greatly appreciated!

stephan-koenig reviewed (changes requested):
Hi Gil,

Great work! So sorry it took me so long!

Not sure if Github saved my old comments, but please rename the file and folder to `tidyverse_ws` for consistent labelling.

- subscripts for O2 etc.
- use dashes for "x-axis", "x-labels" etc.
- Some final exercises would be great. For example, show them a plot without the codechunk (`echo = FALSE` I believe) and ask them to create it.

Best,

Stephan

GilHenriques commented on inst/tutorials/intro_tidyverse/intro_tidyverse.Rmd:
```diff
@@ -0,0 +1,1215 @@
+
+Identifying what type and structure of data you are working with is important because certain functions can only be applied to certain types/structures.
+
+### Subsetting data in base R
```
I could go ahead and remove all references to vector/data subsetting in base R, if you think that makes more sense. 
But to be clear, I should still keep the logical conditions, right? Even if they did the intro R workshop, they may be unfamiliar with some of the functions that return Booleans, such as is.na().

stephan-koenig replied:
Do we subset vectors for any particular use in tidyverse? If not, I would drop all subsetting. Maybe you can move the section as additional exercises in the end?
I would introduce the logical conditions with `filter()` (or do they use it in any other context?).

GilHenriques commented on inst/tutorials/intro_tidyverse/intro_tidyverse.Rmd:
```diff
@@ -0,0 +1,1215 @@
+
+Subsetting by condition is particularly powerful when using conditional statements. Here are some examples of conditional statements:
+
+* `a == b` returns `TRUE` when `a` matches `b`
+* `a != b` returns `TRUE` when `a` does not match `b`
+* `a > b` returns `TRUE` when `a` is greater than `b`
+* `a <= b` returns `TRUE` when `a` is smaller than or equal to `b`
```
I don't know that there is a way to make tables that contain code chunks in them in RMarkdown.

stephan-koenig replied:
For example:

R code     | meaning
---------- | ---------------
`==`       | equals
`< or >`   | less/greater than
`<= or >= `| less/greater than or equal to
`%in%`     | in
`is.na`    | is missing (`NA`)
`!`        | not (as in not equal to `!=`)
`&`        | and
`\|`        | or

GilHenriques commented on inst/tutorials/intro_tidyverse/intro_tidyverse.Rmd:
```diff
@@ -0,0 +1,1215 @@
+
+### Mutate
+
+Mutate is one of the most useful verbs in the dplyr package.
+Use `mutate(y = function(x))` to apply a transformation to some column `x` and assign the output to the column name `y`. 
+The transformation can be a function (e.g., `log(x)`) or a mathematical operation. 
+As an example, run the code below to multiply `Depth` by 1000 (converting its units from kilometers to meters) and assign the output to a new column, `Depth_meters`.
```
Thanks for noticing this! I can just change the exercise to divide by 1000 instead and transform `Depth` from metres to kilometres.

GilHenriques commented:
All changes have been addressed! Some big changes include:

- Child documents for the "how to" and "data description" sections in **both** the Tidyverse and the intro R and Rstudio tutorials
- An exercise at the end
- Removed the indexing section and moved the logical conditions to the `filter()` section

I hope I haven't broken any of the exercises when I moved sections around! Should I do a new pull request?

stephan-koenig commented:
Great! Yes, reissue this pull request if that is possible. I believe there is a re-request button in the reviewer section on the right-hand side. Maybe that is the right spot to do so. I also suggest that the TAs are also going to do a code review. Please assign them as reviewers and let them know to do so by Tuesday, 12 pm.

GilHenriques commented:
Hi @EDUCE-UBC/tas,

I'd be thankful if you could review this tutorial's code by Tuesday, 12 pm. 
It would be great if you could actually do the tutorial yourselves and confirm that all the exercises are working. I moved some paragraphs around and there is a small chance that I might have broken some of the code in the process (not likely, but possible).

Thank you!

r-karimi commented on inst/tutorials/r_and_rstudio_ws/r_and_rstudio_ws.Rmd:
````diff
@@ -300,19 +304,20 @@ As you can see, when you multiply two vectors with the same length, R multiplies
+```
 
-Instead of loading data from your computer, you can also use an online data set: we will use [this one](https://raw.githubusercontent.com/EDUCE-UBC/workshops_access/master/intro_R_2hr/data.csv), which is a `.csv` file hosted on our [GitHub](https://github.com/EDUCE-UBC/workshops_access). Click on the link and examine the format of the data. Notice that the first row consists of column names, that ...
+The [data file](https://raw.githubusercontent.com/EDUCE-UBC/workshops_access/master/intro_R_2hr/data.csv) is a `.csv` file hosted on our [GitHub](https://github.com/EDUCE-UBC/workshops_access). Click on the link and examine the format of the data. Notice that the first row consists of column names, that each column corresponds to a data variable, and that each row corresponds to one observation.
````
Nice job. I'm learning Tensorflow in Python right now and something that's been very handy is examining your data structures often. I'm sure the students will appreciate this.

r-karimi commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
```diff
@@ -0,0 +1,1060 @@
+
+- Create subsets of rows or columns from data frames using dplyr,
+- Change your data frames between wide and long formats,
+- Create various types of graphics,
```
```suggestion
- Create various types of data visualizations,
```

r-karimi commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
````diff
@@ -0,0 +1,1060 @@
+``` {r child = system.file("resources/markdown", "data_description.Rmd", package = "educer")}
+```
+
+During this tutorial, the data are accessible as a data frame called `geochemicals`. In the interactive code box below, use the function `head` to peak at the first few rows of this data frame:
````
```suggestion
During this tutorial, the data are accessible as a data frame called `geochemicals`. In the interactive code box below, use the function `head` to peak at the first five rows of this data frame:
```

r-karimi commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
```diff
@@ -0,0 +1,1060 @@
+The [tidyverse](https://www.tidyverse.org/) is a collection of R packages. Packages expand R's capabilities by providing additional functions. Some tidyverse packages that we will use in this tutorial are: dplyr (for data manipulation; pronounced *dee-plier*), tidyr (for rearranging data frames), and ggplot2 (for data visualization).  We will also use the packages lubridate (for working with date ...
+
+When you work in your own projects, always load the packages you need at the top of your script. You can load each one of the tidyverse packages individually, but since they are often used together, you can also download and load the entire suite of tidyverse packages at once. In the future, you can use the code below to load all of the R packages we will use in this tutorial. 
+
```
```suggestion
Alternatively, when you are working on a large project with version control, you can create R's equivalent of a virtual environment with the `packrat` package, such that anyone who wishes to build upon your code in the future will do so with the same packages and versions you used while developing the project.
```

stephan-koenig replied:
This introduces two new concepts, reproducibility in general and version control specifically. This does not really address the point the paragraph is making (that you have to load packages even if they are installed IMHO). I would not add this suggestion.

r-karimi replied:
Fair enough! I thought the idea of a `virtualenv` might click for people who are used to Python, but I can see how it is extraneous to this tutorial.

r-karimi commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
```diff
@@ -0,0 +1,1060 @@
+
+The tidyverse package magrittr introduced the `%>%` (pipe) operator, which uses this metaphor to make code easier to read:
+
+`f(x) %>% g(y)` is the same as `g(f(x),y)`
```
Nice! This is a very succinct way of explaining piping, and this exact line is what made it make sense to me when I was first learning about the pipe operator. Well done!

r-karimi commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
````diff
@@ -0,0 +1,1060 @@
+               values_to = "Concentration")
+```
+
+We can undo this operation with the opposite function, `pivot_wider`:
````
I know it's similar to `pivot_longer()` but it might be worth explaining what the arguments of `pivot_wider()` actually do as well.

r-karimi commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
```diff
@@ -0,0 +1,1060 @@
+- [ggplot tutorial 1](http://r-statistics.co/Complete-Ggplot2-Tutorial-Part1-With-R-Code.html)
+- [ggplot tutorial 2](https://rpubs.com/g_jw/ggplot2_tutorial)
+- [ggplot tutorial 3](http://tutorials.iq.harvard.edu/R/Rgraphics/Rgraphics.html#the_1_faq)
+- [ggplot2: Elegant Graphics for Data Analysis (online free book)](https://ggplot2-book.org/)
```
Looks great to me! Good work, Gil.

r-karimi reviewed (commented):
Hi Gil, hope these comments are helpful for you.

stephan-koenig commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
```diff
@@ -462,7 +462,14 @@ long_dat <- dat %>%
+We can undo this operation with the opposite function, `pivot_wider()`, which works similarly to its counterpart. The `pivot_wider()` function takes the following arguments:
+
+- The data frame to be pivoted;
+- A vector containing the columns that we want to pivot **or** the columns that we do *not* want to pivot (indicated by a minus sign). If this argument is left blank, it defaults to all columns in the data frame except for those specified in `names_from` and `values_from`;
```
```suggestion
- A vector containing the columns that we want to pivot **or** the columns that we do *not* want to pivot (indicated by an exclamation point `!`). If this argument is left blank, it defaults to all columns in the data frame except for those specified in `names_from` and `values_from`;
```
You new syntax now uses `!` for the anti-(not)-set.

stephan-koenig commented on inst/resources/markdown/how_to_follow_tutorials.Rmd:
```diff
@@ -0,0 +1,9 @@
+
+![](images/code_exercise_small.png){width=5in}
+
+In both text paragraphs and inside interactive code windows, whenever you see any code surrounded by `<angle_brackets>`, that always represents a placeholder, which you should replace with your own code.
```
```suggestion
In both text paragraphs and inside interactive code windows, any placeholder you should replace with your own code is indicated by `<placeholder>`.
```

stephan-koenig commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
````diff
@@ -0,0 +1,1067 @@
+``` {r child = system.file("resources/markdown", "data_description.Rmd", package = "educer")}
+```
+
+During this tutorial, the data are accessible as a data frame called `geochemicals`. In the interactive code box below, use the function `head` to peak at the first five rows of this data frame:
````
```suggestion
During this tutorial, the data are accessible as a data frame called `geochemicals`. In the interactive code box below, use the function `head()` to peak at the first five rows of this data frame:
```

stephan-koenig commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
```diff
@@ -0,0 +1,1067 @@
+- `rename()` columns.
+- `arrange()` a column in ascending or descending order.
+- `mutate()` all values of a column, i.e., add a new column that is a function of existing columns.
+- `group_by()` a variable and `summarise` data by the grouped variable.
```
```suggestion
- `group_by()` a variable and `summarise()` data by the grouped variable.
```

stephan-koenig commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
```diff
@@ -0,0 +1,1067 @@
+`is.na`    | is missing (`NA`)
+`!`        | not (e.g. not equal to `!=`; not missing `!is.na()`)
+`&`        | and
+`\|`        | or
```
```suggestion
`|`        | or
```

stephan-koenig commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
```diff
@@ -0,0 +1,1067 @@
+#### Exercise: Piping
+
+In the box below, rewrite your code from the previous exercise using pipes.
+**Tip:** In RStudio, you can use the keyboard shortcut <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>m</kbd> to enter the pipe operator.
```
```suggestion
**Tip:** In RStudio, you can use the keyboard shortcut <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>M</kbd> to enter the pipe operator.
```

stephan-koenig commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
````diff
@@ -0,0 +1,1067 @@
+long_dat
+```
+
+In the box below, `filter` this data frame so that it includes only oxygen and carbon dioxide. Then, plot the relationship between `Concentration` (in the x-axis) and `Depth` (in the y-axis), indicating the `Chemical` variable by point color:
````
```suggestion
In the box below, `filter()` this data frame so that it includes only oxygen and carbon dioxide. Then, plot the relationship between `Concentration` (in the x-axis) and `Depth` (in the y-axis), indicating the `Chemical` variable by point color:
```

stephan-koenig commented on inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd:
```diff
@@ -0,0 +1,1067 @@
+
+#### Labels
+
+We can change the labels of the previous plot with the `labs` function, specifying exactly what we want as the title, x-, and y-axis labels:
```
```suggestion
We can change the labels of the previous plot with the `labs()` function, specifying exactly what we want as the title, x-, and y-axis labels:
```

stephan-koenig reviewed (dismissed):
Great!
